server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java

```diff
@@ -0,0 +1,137 @@
+                .where(USERNAME.eq(name.asString()))))
+            .map(record -> new DefaultUser(name, record.get(HASHED_PASSWORD),
+                Algorithm.of(record.get(ALGORITHM), fallbackHashingMode), algorithm));
+    }
+
+    @Override
+    public void updateUser(User user) throws UsersRepositoryException {
+        Preconditions.checkArgument(user instanceof DefaultUser);
+        DefaultUser defaultUser = (DefaultUser) user;
+
+        boolean executed = postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+                .set(HASHED_PASSWORD, defaultUser.getHashedPassword())
+                .set(ALGORITHM, defaultUser.getHashAlgorithm().asString())
+                .where(USERNAME.eq(user.getUserName().asString()))
+                .returning(USERNAME)))
+            .map(record -> record.get(USERNAME))
+            .blockOptional()
+            .isPresent();
+
+        if (!executed) {
+            throw new UsersRepositoryException("Unable to update user");
+        }
+    }
+
+    @Override
+    public void removeUser(Username name) throws UsersRepositoryException {
+        boolean executed = postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+                .where(USERNAME.eq(name.asString()))
+                .returning(USERNAME)))
+            .map(record -> record.get(USERNAME))
+            .blockOptional()
+            .isPresent();
+
+        if (!executed) {
+            throw new UsersRepositoryException("Unable to update user");
+        }
+    }
+
+    @Override
+    public boolean contains(Username name) throws UsersRepositoryException {
+        return getUserByName(name).isPresent();
+    }
+
+    @Override
+    public int countUsers() throws UsersRepositoryException {
+        return postgresExecutor.executeRow(dsl -> Mono.from(dsl.select(count()).from(TABLE_NAME)))
+            .map(record -> record.get(0, Integer.class))
+            .block();
+    }
+
+    @Override
+    public Iterator<Username> list() throws UsersRepositoryException {
+        return listReactive()
+            .toIterable()
+            .iterator();
+    }
+
+    @Override
+    public Flux<Username> listReactive() {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.selectFrom(TABLE_NAME)))
+            .map(record -> Username.of(record.get(USERNAME)));
+    }
+
+    @Override
+    public void addUser(Username username, String password) {
+        DefaultUser user = new DefaultUser(username, algorithm, algorithm);
+        user.setPassword(password);
+
+        postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME, USERNAME, HASHED_PASSWORD, ALGORITHM)
+                .values(user.getUserName().asString(), user.getHashedPassword(), user.getHashAlgorithm().asString())))
+            .onErrorMap(e -> e instanceof DataAccessException && e.getMessage().contains(DUPLICATE_VIOLATION_MESSAGE),
```

Comment:
```suggestion
e -> e instanceof DataAccessException && e.getMessage().contains(DUPLICATE_VIOLATION_MESSAGE)
```

Could we have a PG utility wrapping this common tricky condition?

Reply:
Or maybe can we consider it's enough on DataAccessException without having to check the content of the message?

Reply:
But we are not sure whenever get `DataAccessException`, it is exactly a "duplicate data" error.

Reply:
IMO I would be glad to read `onErrorResume(PGPredicate.ALREADY_EXIST_EXCEPTION, e -> new AlreadyExistInUsersRepositoryException("User with username " + username + " already exist!"))` and not get that duplicated everywhere.

Reply:
Hmmm... get your point. So stupidly verbose though :)

Reply:
fixed